flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultPartition.java

```diff
@@ -19,432 +19,105 @@
-				}
-			}
-		}
-	}
-
-	public void destroyBufferPool() {
-		if (bufferPool != null) {
-			bufferPool.lazyDestroy();
-		}
-	}
-
-	/**
-	 * Returns the requested subpartition.
-	 */
-	public ResultSubpartitionView createSubpartitionView(int index, BufferAvailabilityListener availabilityListener) throws IOException {
-		int refCnt = pendingReferences.get();
-
-		checkState(refCnt != -1, "Partition released.");
-		checkState(refCnt > 0, "Partition not pinned.");
-
-		checkElementIndex(index, subpartitions.length, "Subpartition not found.");
-
-		ResultSubpartitionView readView = subpartitions[index].createReadView(availabilityListener);
-
-		LOG.debug("Created {}", readView);
-
-		return readView;
-	}
-
-	public Throwable getFailureCause() {
-		return cause;
+		partitionWriter.flush(subpartitionIndex);
 	}
 
 	@Override
-	public int getNumTargetKeyGroups() {
-		return numTargetKeyGroups;
-	}
-
-	/**
-	 * Releases buffers held by this result partition.
-	 *
-	 * <p>This is a callback from the buffer pool, which is registered for result partitions, which
-	 * are back pressure-free.
-	 */
-	@Override
-	public void releaseMemory(int toRelease) throws IOException {
-		checkArgument(toRelease > 0);
-
-		for (ResultSubpartition subpartition : subpartitions) {
-			toRelease -= subpartition.releaseMemory();
-
-			// Only release as much memory as needed
-			if (toRelease <= 0) {
-				break;
-			}
-		}
-	}
-
-	@Override
-	public String toString() {
-		return "ResultPartition " + partitionId.toString() + " [" + partitionType + ", "
-				+ subpartitions.length + " subpartitions, "
-				+ pendingReferences + " pending references]";
-	}
-
-	// ------------------------------------------------------------------------
-
-	/**
-	 * Pins the result partition.
-	 *
-	 * <p>The partition can only be released after each subpartition has been consumed once per pin
-	 * operation.
-	 */
-	void pin() {
-		while (true) {
-			int refCnt = pendingReferences.get();
-
-			if (refCnt >= 0) {
-				if (pendingReferences.compareAndSet(refCnt, refCnt + subpartitions.length)) {
-					break;
-				}
-			}
-			else {
-				throw new IllegalStateException("Released.");
-			}
-		}
-	}
-
-	/**
-	 * Notification when a subpartition is released.
-	 */
-	void onConsumedSubpartition(int subpartitionIndex) {
-
-		if (isReleased.get()) {
-			return;
-		}
-
-		int refCnt = pendingReferences.decrementAndGet();
-
-		if (refCnt == 0) {
-			partitionManager.onConsumedPartition(this);
-		}
-		else if (refCnt < 0) {
-			throw new IllegalStateException("All references released.");
-		}
-
-		LOG.debug("{}: Received release notification for subpartition {} (reference count now at: {}).",
-				this, subpartitionIndex, pendingReferences);
-	}
-
-	ResultSubpartition[] getAllPartitions() {
-		return subpartitions;
-	}
-
-	// ------------------------------------------------------------------------
+	public void finish() throws IOException {
+		partitionWriter.finish();
```

Comment:
it seems we also need to check success of `partitionWriter.finish()` to preserve the old behaviour.
Maybe, `ResultPartitionWriter.finish()` could also return boolean.

Reply:
Yes, it is better to keep consistent with previous behavior. I would fix it.

Reply:
True, that previous try/finally is not needed actually, it confused me :)